crypto/s2n_rsa_pss.c

```diff
@@ -0,0 +1,221 @@
+    notnull_check(digest_alg);
+
+    GUARD_OSSL(EVP_PKEY_CTX_set_signature_md(ctx, digest_alg), S2N_ERR_INVALID_SIGNATURE_ALGORITHM);
+    GUARD_OSSL(EVP_PKEY_CTX_set_rsa_mgf1_md(ctx, digest_alg), S2N_ERR_INVALID_SIGNATURE_ALGORITHM);
+
+    return 0;
+}
+
+static int s2n_rsa_pss_sign(const struct s2n_pkey *priv, struct s2n_hash_state *digest, struct s2n_blob *signature_out)
+{
+    notnull_check(priv);
+
+    uint8_t digest_length;
+    uint8_t digest_data[S2N_MAX_DIGEST_LEN];
+    GUARD(s2n_hash_digest_size(digest->alg, &digest_length));
+    GUARD(s2n_hash_digest(digest, digest_data, digest_length));
+
+    const EVP_MD* digest_alg = s2n_hash_alg_to_evp_alg(digest->alg);
+    notnull_check(digest_alg);
+
+    /* For more info see: https://www.openssl.org/docs/manmaster/man3/EVP_PKEY_sign.html */
+    EVP_PKEY_CTX *ctx  = EVP_PKEY_CTX_new(priv->key.rsa_pss_key.pkey, NULL);
+    notnull_check(ctx);
+
+    size_t signature_len = signature_out->size;
+    GUARD_OSSL(EVP_PKEY_sign_init(ctx), S2N_ERR_SIGN);
+    GUARD_OSSL(EVP_PKEY_CTX_set_rsa_padding(ctx, RSA_PKCS1_PSS_PADDING), S2N_ERR_SIGN);
+    GUARD(s2n_evp_ctx_set_signature_digest(ctx, digest_alg, S2N_ERR_SIGN));
+    GUARD_OSSL(EVP_PKEY_CTX_set_rsa_pss_saltlen(ctx, RSA_PSS_SALTLEN_DIGEST), S2N_ERR_SIGN);
+
+    /* Calling EVP_PKEY_sign() with NULL will only update the signature_len parameter so users can validate sizes. */
+    GUARD_OSSL(EVP_PKEY_sign(ctx, NULL, &signature_len, digest_data, digest_length), S2N_ERR_SIGN);
+    S2N_ERROR_IF(signature_len > signature_out->size, S2N_ERR_SIZE_MISMATCH);
+
+    /* Actually sign the the digest */
+    GUARD_OSSL(EVP_PKEY_sign(ctx, signature_out->data, &signature_len, digest_data, digest_length), S2N_ERR_SIGN);
+    signature_out->size = signature_len;
+
+    return 0;
+}
+
+static int s2n_rsa_pss_verify(const struct s2n_pkey *pub, struct s2n_hash_state *digest, struct s2n_blob *signature_in)
+{
+    notnull_check(pub);
+
+    uint8_t digest_length;
+    uint8_t digest_data[S2N_MAX_DIGEST_LEN];
+    GUARD(s2n_hash_digest_size(digest->alg, &digest_length));
+    GUARD(s2n_hash_digest(digest, digest_data, digest_length));
+    const EVP_MD* digest_alg = s2n_hash_alg_to_evp_alg(digest->alg);
+
+    /* For more info see: https://www.openssl.org/docs/manmaster/man3/EVP_PKEY_verify.html */
+    EVP_PKEY_CTX *ctx = EVP_PKEY_CTX_new(pub->key.rsa_pss_key.pkey, NULL);
+    notnull_check(ctx);
+
+    GUARD_OSSL(EVP_PKEY_verify_init(ctx), S2N_ERR_VERIFY_SIGNATURE);
+    GUARD_OSSL(EVP_PKEY_CTX_set_rsa_padding(ctx, RSA_PKCS1_PSS_PADDING), S2N_ERR_SIGN);
+    GUARD(s2n_evp_ctx_set_signature_digest(ctx, digest_alg, S2N_ERR_VERIFY_SIGNATURE));
+    GUARD_OSSL(EVP_PKEY_verify(ctx, signature_in->data, signature_in->size, digest_data, digest_length), S2N_ERR_VERIFY_SIGNATURE);
+
+    return 0;
+}
+
+static int s2n_rsa_pss_keys_match(const struct s2n_pkey *pub, const struct s2n_pkey *priv)
```

Comment:
What is the purpose of this method?  At the very least you should also check that the modulus for both the public and private keys are the same.

Even more cautious (and likely unnecessary) checks are as follows:
* Ensure the public exponent (if present) on the private key matches the public exponent on the private key
* Ensure the primes (if present) on the private key multiply to be equal to the modulus
* Ensure the 3 CRT parameters (imp1, imq1, coef, if I recall the names correctly) are correct.

Reply:
It's called once at startup just after loading the Certificate and Private Key to verify that they are a match. It's solely used to detect server-side misconfigurations (Eg someone updates the Certificate, but not the Private Key, or vice-versa).

Reply:
Fixed